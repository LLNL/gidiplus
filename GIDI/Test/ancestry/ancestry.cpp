/*
# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>
*/

#include <stdlib.h>
#include <iostream>

#include "GIDI_testUtilities.hpp"

static char const *description = "This code checks that the Ancestry class functions findInAncestry and toXLink are working, and that GIDI set setAncestry in all the required objects.";

void main2( int argc, char **argv );
void printReaction( GIDI::Reaction *a_reaction, GIDI::Ancestry *a_protare );
void printSuite( GIDI::Suite &a_suite, GIDI::Ancestry *a_protare, GIDI::Ancestry *a_reaction, bool a_isProduct = false );
void printOutputChannel( GIDI::OutputChannel *a_outputChannel, GIDI::Ancestry *a_protare, GIDI::Ancestry *a_reaction );
std::string checkXLink( GIDI::Ancestry &ancestry, GIDI::Ancestry *a_protare, GIDI::Ancestry *a_reaction );
void checkXLink2( GIDI::Ancestry &a_ancestry, std::string &xlink, GIDI::Ancestry *from );
/*
=========================================================
*/
int main( int argc, char **argv ) {

    try {
        main2( argc, argv ); }
     catch (std::exception &exception) {
        std::cerr << exception.what( ) << std::endl;
        exit( EXIT_FAILURE ); }
    catch (char const *str) {
        std::cerr << str << std::endl;
        exit( EXIT_FAILURE ); }
    catch (std::string &str) {
        std::cerr << str << std::endl;
        exit( EXIT_FAILURE );
    }

    exit( EXIT_SUCCESS );
}
/*
=========================================================
*/
void main2( int argc, char **argv ) {

    argvOptions argv_options( "ancestry", description );
    ParseTestOptions parseTestOptions( argv_options, argc, argv );

    parseTestOptions.m_askGNDS_File = true;

    parseTestOptions.parse( );

    GIDI::Construction::PhotoMode photo_mode = parseTestOptions.photonMode( GIDI::Construction::PhotoMode::atomicOnly );
    GIDI::Construction::Settings construction( GIDI::Construction::ParseMode::all, photo_mode );
    PoPI::Database pops;
    GIDI::Protare *protare = parseTestOptions.protare( pops, "../pops.xml", "../all.map", construction, PoPI::IDs::photon, "O" );

    std::cout << stripDirectoryBase( protare->fileName( ), "/GIDI/Test/" ) << std::endl;

    GIDI::Reaction *reaction = protare->reaction( 0 );
    std::cout << checkXLink( *protare, protare, reaction ) << std::endl;

    printSuite( protare->styles( ), protare, reaction );

    for( std::size_t index = 0; index < protare->numberOfReactions( ); ++index ) {
        GIDI::Reaction *reaction = protare->reaction( index );
        printReaction( reaction, protare );
    }

    for( std::size_t index = 0; index < protare->numberOfOrphanProducts( ); ++index ) {
        GIDI::Reaction *reaction = protare->orphanProduct( index );
        printReaction( reaction, protare );
    }

    delete protare;
}
/*
=========================================================
*/
void printReaction( GIDI::Reaction *a_reaction, GIDI::Ancestry *a_protare ) {

    std::cout << "" << checkXLink( *a_reaction, a_protare, a_reaction ) << std::endl;

    printSuite( a_reaction->doubleDifferentialCrossSection( ), a_protare, a_reaction );
    printSuite( a_reaction->crossSection( ), a_protare, a_reaction );
    printSuite( a_reaction->availableEnergy( ), a_protare, a_reaction );
    printSuite( a_reaction->availableMomentum( ), a_protare, a_reaction );

    printOutputChannel( a_reaction->outputChannel( ), a_protare, a_reaction );
}
/*
=========================================================
*/
void printSuite( GIDI::Suite &a_suite, GIDI::Ancestry *a_protare, GIDI::Ancestry *a_reaction, bool a_isProduct ) {

    std::cout << "    " << checkXLink( a_suite, a_protare, a_reaction ) << std::endl;
    for( std::size_t i1 = 0; i1 < a_suite.size( ); ++i1 ) {
        GIDI::Form *form = a_suite.get<GIDI::Form>( i1 );
        std::cout << "    " << checkXLink( *form, a_protare, a_reaction ) << std::endl;

        if( a_isProduct ) {
            GIDI::Product *product = static_cast<GIDI::Product *>( form );

                printSuite( product->multiplicity( ), a_protare, a_reaction );
                printSuite( product->distribution( ), a_protare, a_reaction );
                printSuite( product->averageEnergy( ), a_protare, a_reaction );
                printSuite( product->averageMomentum( ), a_protare, a_reaction );

                GIDI::OutputChannel *outputChannel = product->outputChannel( );
                if( outputChannel != nullptr ) printOutputChannel( outputChannel, a_protare, a_reaction );
        }
    }
}
/*
=========================================================
*/
void printOutputChannel( GIDI::OutputChannel *a_outputChannel, GIDI::Ancestry *a_protare, GIDI::Ancestry *a_reaction ) {

    std::cout << "    " << checkXLink( *a_outputChannel, a_protare, a_reaction ) << std::endl;
    printSuite( a_outputChannel->Q( ), a_protare, a_reaction );
    printSuite( a_outputChannel->products( ), a_protare, a_reaction, true );
}
/*
=========================================================
*/
std::string checkXLink( GIDI::Ancestry &a_ancestry, GIDI::Ancestry *a_protare, GIDI::Ancestry *a_reaction ) {

    std::string xlink = a_ancestry.toXLink( );

    checkXLink2( a_ancestry, xlink, a_protare );
    checkXLink2( a_ancestry, xlink, a_reaction );

    return( xlink );
}
/*
=========================================================
*/
void checkXLink2( GIDI::Ancestry &a_ancestry, std::string &a_xlink, GIDI::Ancestry *a_from ) {

    GIDI::Ancestry *link = a_from->findInAncestry( a_xlink );

    if( link == nullptr ) {
        std::cout << " LINK NOT FOUND "; }
    else {
        if( &a_ancestry != link ) {
            std::cout << " a_ancestry DOES NOT MATCH link "; }
        else {
            std::string toXLink( link->toXLink( ) );
            if( a_xlink != toXLink ) {
                std::cout << " XLINK DOES NOT MATCH link->toXLink( ) ";
            }
        }
    }
}
