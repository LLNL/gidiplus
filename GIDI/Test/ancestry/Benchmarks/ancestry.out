/.../GIDI/Test/Data/MG_MC/photo-atomic/photoat-008_O_000.xml
/reactionSuite
    /reactionSuite/styles
    /reactionSuite/styles/evaluated[@label='eval']
    /reactionSuite/styles/averageProductData[@label='apd']
    /reactionSuite/styles/MonteCarlo_cdf[@label='MonteCarlo_cdf']
    /reactionSuite/styles/multiGroup[@label='MultiGroup']
    /reactionSuite/styles/heated[@label='heated_000']
    /reactionSuite/styles/griddedCrossSection[@label='MonteCarlo_000']
    /reactionSuite/styles/heatedMultiGroup[@label='MultiGroup_000']
/reactionSuite/reactions/reaction[@label='O + photon [coherent]']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/doubleDifferentialCrossSection
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/doubleDifferentialCrossSection/coherentPhotonScattering[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/crossSection
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/crossSection/XYs1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/crossSection/Ys1d[@label='MonteCarlo_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/crossSection/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/availableEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/availableEnergy/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/availableEnergy/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/availableMomentum
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/availableMomentum/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/availableMomentum/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/Q
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/Q/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/Q/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/multiplicity
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/multiplicity/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/distribution
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/distribution/coherentPhotonScattering[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/distribution/multiGroup3d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/averageProductEnergy/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/averageProductEnergy/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/averageProductMomentum
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/averageProductMomentum/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='photon']/averageProductMomentum/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']/multiplicity
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']/distribution
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']/distribution/unspecified[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [coherent]']/outputChannel/products/product[@label='O']/averageProductMomentum
/reactionSuite/reactions/reaction[@label='O + photon [incoherent]']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/doubleDifferentialCrossSection
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/doubleDifferentialCrossSection/incoherentPhotonScattering[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/crossSection
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/crossSection/XYs1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/crossSection/Ys1d[@label='MonteCarlo_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/crossSection/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/availableEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/availableEnergy/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/availableEnergy/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/availableMomentum
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/availableMomentum/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/availableMomentum/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/Q
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/Q/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/Q/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/multiplicity
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/multiplicity/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/distribution
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/distribution/incoherentPhotonScattering[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/distribution/multiGroup3d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/averageProductEnergy/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/averageProductEnergy/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/averageProductMomentum
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/averageProductMomentum/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='photon']/averageProductMomentum/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']/multiplicity
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']/distribution
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']/distribution/unspecified[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [incoherent]']/outputChannel/products/product[@label='O']/averageProductMomentum
/reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/doubleDifferentialCrossSection
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/crossSection
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/crossSection/regions1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/crossSection/XYs1d[@label='heated_000']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/crossSection/Ys1d[@label='MonteCarlo_000']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/crossSection/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/availableEnergy
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/availableEnergy/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/availableEnergy/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/availableMomentum
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/availableMomentum/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/availableMomentum/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/Q
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/Q/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/Q/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']/multiplicity
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']/distribution
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']/distribution/unspecified[@label='eval']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='e-']/averageProductMomentum
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']/multiplicity
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']/distribution
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']/distribution/unspecified[@label='eval']
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='e- + O [photo-electric]']/outputChannel/products/product[@label='O']/averageProductMomentum
/reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/doubleDifferentialCrossSection
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/crossSection
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/crossSection/regions1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/crossSection/XYs1d[@label='heated_000']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/crossSection/Ys1d[@label='MonteCarlo_000']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/crossSection/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/availableEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/availableEnergy/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/availableEnergy/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/availableMomentum
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/availableMomentum/XYs1d[@label='apd']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/availableMomentum/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/Q
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/Q/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/Q/gridded1d[@label='MultiGroup_000']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']/multiplicity
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']/distribution
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']/distribution/unspecified[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='photon']/averageProductMomentum
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']/multiplicity
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']/multiplicity/constant1d[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']/distribution
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']/distribution/unspecified[@label='eval']
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']/averageProductEnergy
    /reactionSuite/reactions/reaction[@label='O + photon [pair production: electron field]']/outputChannel/products/product[@label='O']/averageProductMomentum
